### ISSUE TITLE
A strange gpu memory usage increase in training process

### ISSUE THREAD

**Issue by jiaerfei** [ultralytics/ultralytics#19862]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi, thank you for the excellent work !
I meet an issue when training yolov8m, as shown in the figure, the gpu usage increases with the iterative training process, I am quite confused, the code is just like the template, thank you for giving some advice on this problem.
Thank you for your reply !

![Image](https://github.com/user-attachments/assets/5739ba74-e91b-4e5d-bb00-df97b8765208)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @jiaerfei, thank you for bringing this to our attention and for your interest in Ultralytics 🚀! We appreciate your support and detailed report.

For GPU memory usage issues, we recommend starting by reviewing the [Docs](https://docs.ultralytics.com/) for best practices and tips. If possible, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us better understand and debug the issue. This should include the specific training command used, any custom modifications, and, if applicable, the dataset details.

### Upgrade
Ensure you're using the latest version of the `ultralytics` package, as your issue may already be resolved. Upgrade with:
```bash
pip install -U ultralytics
```

### Environments
YOLO can be run in the following verified environments:
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud**: Deep Learning VM ([GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/))
- **Amazon**: Deep Learning AMI ([AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/))
- **Docker**: ([Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/)) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

### Community Support
If you'd like to discuss this further with the community or get real-time help:
- Join us on [Discord](https://discord.com/invite/ultralytics) 🎧
- Visit our [Discourse](https://community.ultralytics.com/) for detailed discussions
- Engage with others on our [Subreddit](https://reddit.com/r/Ultralytics)

### Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

Our CI tests verify all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) across macOS, Windows, and Ubuntu. If the badge is green, all tests are passing.

This is an automated response to help guide you, but rest assured an Ultralytics engineer will review your issue and assist you soon. Thank you for your patience! 😊

**Comment by Y-T-G** [comment]
Can you provide the output after running this command in terminal: `yolo checks`?

**Comment by jiaerfei** [comment]
> Can you provide the output after running this command in terminal: `yolo checks`?
Yes, this is the output.
Besides, I found that the increase issue occur when I firstly running the code on a dataset, if a label cache is created, then in the following training, gpu usage increases, but if I stop this training and restart it, the gpu usage can keep a stable value.

![Image](https://github.com/user-attachments/assets/601372d8-4ced-4c81-9fc4-c3130963c280)

**Comment by jiaerfei** [comment]
well, maybe it is not the label cache, when I change the model, such as from yolom to yolon, the gpu still increases...

**Comment by jiaerfei** [comment]
It is also observed that 24 workers are created in the training process. I guess that 8 in train_loader, 8 in val_loader, 8 in label_cache, but it is strange that after the label cache is created, everytime I start the training, the worker num is still 24. I don't know if this is normal.

![Image](https://github.com/user-attachments/assets/cf3170d8-f3a7-4185-ad3f-ad20c8877548)

**Comment by Y-T-G** [comment]
Train loader uses twice the workers. So 16 + 8.

**Comment by jiaerfei** [comment]
Thank you for the reply, and do you have any advice on the gpu increase issue?, like what can I do to figure out the cause, I even want to to reinstall the windows system, Thank you

**Comment by Y-T-G** [comment]
Does it cause outofmemoryerror?

**Comment by jiaerfei** [comment]
no, because I will manually stop the training. After some experiments, I found when I firstly create label cache, the following training appear gpu increase. After the label cache is created, if I change model or restart training, gpu increase may disappear.

**Comment by glenn-jocher** [comment]
Thanks for sharing your insights! For users experiencing GPU memory fluctuations, we recommend trying training with `augment=False` and `multi_scale=False` to reduce memory variability, as these augmentations can dynamically impact VRAM usage. If issues persist, consider adjusting your batch size as shown in our Training Tips guide. For more optimization strategies, see our Common Issues documentation.

**Comment by jiaerfei** [comment]
Thank you for the reply!
But here GPU memory is not fluctuated, it just increases with the training, and it only happens when I firstly create label cache, then the following training will appear gpu increase, and it doesn't fluctuated. Here is my 100-epoch training process, almost doubled. So I am quitely confused why the label cache can cause this issue.
Maybe I think it is my windows system problem...

![Image](https://github.com/user-attachments/assets/87010877-8cb3-484b-b610-185a94f6b448)

![Image](https://github.com/user-attachments/assets/5926502f-e3ab-4b7a-86b5-3485c62f5386)

**Comment by Y-T-G** [comment]
What's your training command?

**Comment by jiaerfei** [comment]
This is my training command in Pycharm, here it is opened in windows notebook.
<img width="557" alt="Image" src="https://github.com/user-attachments/assets/84a4aa58-f540-4617-9f52-64ac99f48351" />

**Comment by gboeer** [comment]
Interesting.
It's quite unlikely that this is caused by your Windows installation.
Just to be clear, this only happens the first time you start a training with new data, i.e. when the label cache has to be created?
This does not happen if the label cache is reused ?

Also, did you try to set multi_scale=False  (in the train() function call) just to see if it has an effect?

Another thing to narrow down the problem:
Can you run the script outside PyCharm, i mean just open a Windows cmd and run your script from there.
Just to check if maybe pycharm has a aching / memory issue .. but its just a wild guess.

**Comment by glenn-jocher** [comment]
@gboeer thank you for sharing these details. Based on your observations, this appears related to the label cache initialization process during first-run training. For investigation:

1. Try adding `multi_scale=False` to your training arguments as shown in the [Train Settings documentation](https://docs.ultralytics.com/modes/train/) to eliminate scaling-related memory variations  
2. As suggested, test running directly from Windows CMD (outside PyCharm) to isolate any IDE-specific memory management behaviors  
3. Consider setting explicit `cache=False` in your training command to compare memory patterns  

These steps will help narrow down whether it's cache-related allocation or environment-specific behavior. The [Training Tips guide](https://docs.ultralytics.com/guides/model-training-tips/) also contains useful optimization strategies for memory management.

**Comment by jiaerfei** [comment]
@gboeer @glenn-jocher Thank you for your advice! I have tried the settings and environment, but gpu increase still appear when I firstly create label cache, and it disappear when I reuse the label cache.
Besides, I found gpu increase in my computer mainly has three properties, 
1. label cache is firstly created. 
2. it happens at the end of each epoch, there is a sudden gpu increase.
3. The number of gpu increase is not too large, maybe 0.02~0.04GB/epoch, a little different for n/s/m/l
I have also reinstall the windows system but it doesn't disappear, later I will try it on other computers.

**Comment by jiaerfei** [comment]
Got it. Thank you for the reply again! I will try these settings.

**Comment by glenn-jocher** [comment]
You're welcome! Please let us know if adjusting the `multi_scale` and `workers` settings resolves the memory pattern you're observing. If the behavior persists, we recommend checking our [Common Issues guide](https://docs.ultralytics.com/guides/yolo-common-issues/) for additional troubleshooting steps. Happy training! 🚀

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in https://github.com/ultralytics/ultralytics/pull/20179, thanks to the efforts of @Y-T-G with contributions from @Laughing-q and @glenn-jocher.

The key change in this update is a reduction in the GPU memory usage threshold that triggers memory clearing during training—from 90% down to 50%. This improvement aims to free up GPU resources earlier, which should help prevent out-of-memory crashes and ensure smoother training, especially on systems with limited memory.

To test if this resolves the issue, try installing the latest codebase with:
```
pip install git+https://github.com/ultralytics/ultralytics.git@main
```
Alternatively, you can wait for the next official PyPI release, which will include this change.

Please let us know if this update resolves your problem or if you encounter any further issues. Thank you for bringing this to our attention! Your feedback is always appreciated.

**Comment by glenn-jocher** [comment]
@jiaerfei The memory increase pattern you're seeing appears to be addressed in the recent fix mentioned by UltralyticsAssistant (PR #20179). The change lowers the memory threshold for GPU clearing from 90% to 50%, which should help with the gradual accumulation you observed.

I'd recommend installing the updated version as suggested to test if this resolves your issue. If you still experience memory increases after the update, try reducing workers further and setting `cache='disk'` as previously suggested. 

Your detailed observations about this behavior being connected to label cache creation have been valuable for our team to identify this pattern. We'll continue monitoring Windows training performance to ensure consistent memory management.
